### PR TITLE
[Build] Add an option to emit Swift module separately

### PR DESCRIPTION
**Pull request by aciidgh**

### CODE REVIEW

File: Sources/Build/BuildPlan.swift [hunk 1]
```diff
@@ -701,11 +708,148 @@ public final class SwiftTargetBuildDescription {
...
+        // FIXME: Handle WMO
+
+        for source in target.sources.paths {
+            result.append(source.pathString)
+        }
+
+        result.append("-I")
+        result.append(buildParameters.buildPath.pathString)
+
+        // FIXME: Maybe refactor these into "common args".
```
Comment (DougGregor) [inline Sources/Build/BuildPlan.swift]: Yes please :)

File: Sources/Build/BuildPlan.swift [hunk 2]
```diff
@@ -701,11 +708,148 @@ public final class SwiftTargetBuildDescription {
...
+        result.append("-emit-module")
+        result.append("-emit-module-path")
+        result.append(moduleOutputPath.pathString)
+        result += buildParameters.toolchain.extraSwiftCFlags
+
+        result.append("-Xfrontend")
+        result.append("-experimental-skip-non-inlinable-function-bodies")
+        result.append("-force-single-frontend-invocation")
+
+        if target.type == .library || target.type == .test {
```
Comment (DougGregor) [inline Sources/Build/BuildPlan.swift]: Factor out or turn into a `switch` to catch changes to target types?

**Comment by harlanhaskins** [comment]
:shipit:

File: Sources/Build/BuildPlan.swift [hunk 3]
```diff
@@ -701,11 +708,148 @@ public final class SwiftTargetBuildDescription {
...
+
+            let sourceFileName = source.basenameWithoutExt
+            let partialModulePath = objectDir.appending(component: sourceFileName + "~partial.swiftmodule")
+
+            let swiftDepsPath = objectDir.appending(component: sourceFileName + ".swiftdeps")
+
+            stream <<< "  \"" <<< source.pathString <<< "\": {\n"
+            // FIXME: Handle WMO
+            let depsPath = objectDir.appending(component: sourceFileName + ".d")
+            stream <<< "    \"dependencies\": \"" <<< depsPath.pathString <<< "\",\n"
```
Comment (DougGregor) [inline Sources/Build/BuildPlan.swift]: There's no existing abstraction for emitting the output file map in SwiftPM? Should we sink down [`OutputFileMap`](https://github.com/apple/swift-driver/blob/master/Sources/SwiftDriver/Driver/OutputFileMap.swift) from the Swift driver?

File: Sources/Build/ManifestBuilder.swift [hunk 4]
```diff
@@ -149,6 +149,78 @@ extension LLBuildManifestBuilder {
...
+            args: target.emitObjectsCommandLine()
+        )
+    }
+
+    private func addCmdWithBuiltinSwiftTool(
+        _ target: SwiftTargetBuildDescription,
+        inputs: [Node],
+        cmdOutputs: [Node]
+    ) {
+        let isLibrary = target.target.type == .library || target.target.type == .test
```
Comment (DougGregor) [inline Sources/Build/ManifestBuilder.swift]: Can we make this a query on `target.target.type`? (Can you tell I'm not a fan of checking specific cases on enums?)

File: Sources/Build/BuildPlan.swift (hunk 3 above)
Comment (aciidb0mb3r) [reply Sources/Build/BuildPlan.swift]: Nope. Yeah, maybe we can bring that type in from the driver.

**Pull request merged**